service/lister/service.go

```diff
@@ -0,0 +1,110 @@
+// Package lister offers a service to retrieve credentials for an organization.
+package lister
+
+import (
+	"fmt"
+	"strings"
+
+	"github.com/giantswarm/microerror"
+	"github.com/giantswarm/micrologger"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/client-go/kubernetes"
+)
+
+const (
+	// the namespace in which we store credentiald secrets
+	kubernetesCredentialNamespace = "giantswarm"
+
+	// the selector we use to retrieve credentials
+	kubernetesLabelSelectorMask = "app=credentiald,giantswarm.io/organization=%s"
+
+	gaugeValue = float64(1)
+
+	providerAWS   = "aws"
+	providerAzure = "azure"
+
+	// We use these data keys to detect the provider from a secret.
+	providerAWSDetectionKey   = "aws.admin.arn"
+	providerAzureDetectionKey = "azure.azureoperator.subscriptionid"
+)
+
+// Config is the service configuration data structure.
+type Config struct {
+	K8sClient kubernetes.Interface
+	Logger    micrologger.Logger
+}
+
+// Service is our actual service.
+type Service struct {
+	k8sClient kubernetes.Interface
+	logger    micrologger.Logger
+}
+
+// New creates a new lister service based on a configutation.
+func New(config Config) (*Service, error) {
+	if config.K8sClient == nil {
+		return nil, microerror.Maskf(invalidConfigError, "%T.K8sClient must not be empty", config)
+	}
+
+	if config.Logger == nil {
+		return nil, microerror.Maskf(invalidConfigError, "%T.Logger must not be empty", config)
+	}
+
+	service := &Service{
+		k8sClient: config.K8sClient,
+		logger:    config.Logger,
+	}
+
+	return service, nil
+}
+
+// List returns metadata on all credentials found.
+func (c *Service) List(request Request) ([]*Response, error) {
+	c.logger.Log("level", "debug", "message", fmt.Sprintf("listing secrets for organization %s", request.Organization))
+
+	selector := fmt.Sprintf(kubernetesLabelSelectorMask, request.Organization)
+	credentialList, err := c.k8sClient.CoreV1().Secrets(kubernetesCredentialNamespace).List(metav1.ListOptions{
+		LabelSelector: selector,
+	})
+	if err != nil {
+		c.logger.Log("level", "error", "message", "could not list secrets", "stack", fmt.Sprintf("%#v", err))
+	}
```

Comment:
@kopiczko I have second thoughts here. If we can't list secrets for some reasons, wouldn't that be worth an error 500?

Reply:
100%. I missed that.